lib/async_hooks.js

```diff
@@ -211,19 +212,73 @@ class AsyncResource {
 }
 
 const storageList = [];
+const seenLayer = [];
+let trackerCount = 0;
+let depth = 0;
+
+function refreshStorageHooks() {
+  if (storageList.length === 0) {
+    storageHookWithTracking.disable();
+    storageHook.disable();
+  } else if (trackerCount > 0) {
+    storageHookWithTracking.enable();
+    storageHook.disable();
+  } else {
+    storageHookWithTracking.disable();
+    storageHook.enable();
+  }
+}
+
+function patchPromiseBarrier(currentResource) {
+  PromiseResolve({
+    then(resolve) {
+      const resource = executionAsyncResource();
+      propagateToStorageLists(resource, currentResource);
+      resolve();
+    }
+  });
+}
+
+function propagateToStorageLists(resource, currentResource) {
+  for (let i = 0; i < storageList.length; ++i) {
+    storageList[i]._propagate(resource, currentResource);
+  }
+}
+
 const storageHook = createHook({
   init(asyncId, type, triggerAsyncId, resource) {
     const currentResource = executionAsyncResource();
     // Value of currentResource is always a non null object
-    for (let i = 0; i < storageList.length; ++i) {
-      storageList[i]._propagate(resource, currentResource);
+    propagateToStorageLists(resource, currentResource);
+  }
+});
+
+const storageHookWithTracking = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    const currentResource = executionAsyncResource();
+    // Value of currentResource is always a non null object
+    propagateToStorageLists(resource, currentResource);
+
+    if (type === 'PROMISE' && !seenLayer[depth]) {
+      seenLayer[depth] = true;
+      patchPromiseBarrier(currentResource);
     }
+  },
+
+  before(asyncId) {
+    depth++;
+    seenLayer[depth] = false;
+  },
+
+  after(asyncId) {
+    depth--;
   }
 });
 
 class AsyncLocalStorage {
-  constructor() {
+  constructor({ trackAsyncAwait = false } = {}) {
```

Comment:
Not blocking comment:  I'm not a big fan of user facing options which could be even named `disableProblems`...
Anyhow, some doc update is needed. And maybe we should use a more clear name as async await is tracked in general.